workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourLibrary.cs

```diff
@@ -0,0 +1,166 @@
+            "this is invalid and will not be enabled!";
+
+        public SpatialOSBehaviourLibrary(ILogDispatcher logger)
+        {
+            this.logger = logger;
+        }
+
+        public void InjectAllReadersWriters(MonoBehaviour spatialOSBehaviour)
+        {
+            EnsureLoaded(spatialOSBehaviour.GetType());
+            if (badTypes.Contains(spatialOSBehaviour.GetType()))
+            {
+                return;
+            }
+
+            foreach (var readerWriterComponentId in componentReaderIdsForBehaviours[spatialOSBehaviour.GetType()])
+            {
+                Inject(spatialOSBehaviour, readerWriterComponentId);
+            }
+
+            foreach (var readerWriterComponentId in componentWriterIdsForBehaviours[spatialOSBehaviour.GetType()])
+            {
+                Inject(spatialOSBehaviour, readerWriterComponentId);
+            }
+        }
+
+        public void DeInjectAllReadersWriters(MonoBehaviour spatialOSBehaviour)
+        {
+            EnsureLoaded(spatialOSBehaviour.GetType());
+            if (badTypes.Contains(spatialOSBehaviour.GetType()))
+            {
+                return;
+            }
+
+            foreach (var readerWriterComponentId in componentReaderIdsForBehaviours[spatialOSBehaviour.GetType()])
+            {
+                DeInject(spatialOSBehaviour, readerWriterComponentId);
+            }
+
+            foreach (var readerWriterComponentId in componentWriterIdsForBehaviours[spatialOSBehaviour.GetType()])
+            {
+                DeInject(spatialOSBehaviour, readerWriterComponentId);
+            }
+        }
+
+        public IEnumerable<uint> GetRequiredReaderComponentIds(Type behaviourType)
+        {
+            EnsureLoaded(behaviourType);
+            return componentReaderIdsForBehaviours[behaviourType];
+        }
+
+        public IEnumerable<uint> GetRequiredWriterComponentIds(Type behaviourType)
+        {
+            EnsureLoaded(behaviourType);
+            return componentWriterIdsForBehaviours[behaviourType];
+        }
+
+        private void Inject(MonoBehaviour spatialOSBehaviour, uint componentId)
+        {
+            var readerWriter = ReaderWriterFactory.CreateReaderWriter(componentId);
+            var memberAdapter = adapterCache[spatialOSBehaviour.GetType()][componentId];
+            memberAdapter.SetValue(spatialOSBehaviour, readerWriter);
+        }
+
+        private void DeInject(MonoBehaviour spatialOSBehaviour, uint componentId)
+        {
+            var memberAdapter = adapterCache[spatialOSBehaviour.GetType()][componentId];
+            memberAdapter.SetValue(spatialOSBehaviour, null);
+        }
+
+        private void EnsureLoaded(Type behaviourType)
+        {
+            if (adapterCache.ContainsKey(behaviourType))
+            {
+                return;
+            }
+
+            var adapters = reflectionUtil.GetMembersWithMatchingAttributes(behaviourType);
```

Comment:
we can probably just pull this out into a function. This class is mostly just a wrapper around a single function and that function doesn't have all that much code in ti.